gql/state.go

```diff
@@ -131,12 +140,128 @@ func lexInside(l *lex.Lexer) lex.StateFn {
 			return lexArgInside
 		case r == ':':
 			return lexAlias
+		case r == '@':
+			return lexDirective
 		default:
 			return l.Errorf("Unrecognized character in lexInside: %#U", r)
 		}
 	}
 }
 
+// lexFilterFuncInside expects input to look like ("...", "...").
+func lexFilterFuncInside(l *lex.Lexer) lex.StateFn {
+	l.AcceptRun(isSpace)
+	l.Ignore() // Any spaces encountered.
+
+	for {
+		r := l.Next()
+		if isSpace(r) || r == ',' {
+			l.Ignore()
+		} else if r == leftRound {
+			l.Emit(itemLeftRound)
+		} else if r == rightRound {
+			l.Emit(itemRightRound)
+			return lexFilterInside
+		} else if isEndLiteral(r) {
+			l.Ignore()
+			l.AcceptUntil(isEndLiteral) // This call will backup the ending ".
+			l.Emit(itemFilterFuncArg)
+			l.Next() // Consume " and ignore it.
+			l.Ignore()
+		} else {
+			return l.Errorf("Expected quotation mark in lexFilterFuncArgs")
+		}
+	}
+}
+
+// lexFilterFuncName expects input to look like equal("...", "...").
+func lexFilterFuncName(l *lex.Lexer) lex.StateFn {
+	for {
+		// The caller already checked isNameBegin, and absorbed one rune.
+		r := l.Next()
+		if isNameSuffix(r) {
+			continue
+		}
+		l.Backup()
+		l.Emit(itemFilterFunc)
+		break
+	}
+	return lexFilterFuncInside
+}
+
+// lexFilterInside expects input like (  (equal(...) && f(...)) || g(...)  ).
+func lexFilterInside(l *lex.Lexer) lex.StateFn {
+	for {
+		r := l.Next()
+		switch {
+		case r == leftRound:
+			l.Emit(itemLeftRound)
+			l.FilterDepth++
+		case r == rightRound:
+			if l.FilterDepth == 0 {
+				return l.Errorf("Unexpected )")
+			}
+			l.FilterDepth--
+			l.Emit(itemRightRound)
+			if l.FilterDepth == 0 {
+				return lexInside // Filter directive is done.
+			}
+		case r == lex.EOF:
+			return l.Errorf("Unclosed directive")
+		case isSpace(r) || isEndOfLine(r):
+			l.Ignore()
+		case isNameBegin(r):
+			return lexFilterFuncName
+		case r == '&':
+			r2 := l.Next()
+			if r2 == '&' {
+				l.Emit(itemFilterAnd)
+				return lexFilterInside
+			}
+			return l.Errorf("Expected & but got %v", r2)
+		case r == '|':
+			r2 := l.Next()
+			if r2 == '|' {
+				l.Emit(itemFilterOr)
+				return lexFilterInside
+			}
+			return l.Errorf("Expected | but got %v", r2)
+		default:
+			return l.Errorf("Unrecognized character in lexInside: %#U", r)
+		}
+	}
+}
+
+// lexDirective is called right after we see a @.
+func lexDirective(l *lex.Lexer) lex.StateFn {
+	r := l.Next()
+	if !isNameBegin(r) {
+		return l.Errorf("Unrecognized character in lexDirective: %#U", r)
+	}
+
+	l.Backup()
+	buf := bytes.NewBuffer(make([]byte, 0, 15))
```

Comment:
Add a comment about why this buf wouldn't exceed 15 bytes (or can it resize the underlying slice automatically?)